test/torchaudio_unittest/functional/autograd_impl.py

```diff
@@ -250,6 +252,59 @@ def test_bandreject_biquad(self, central_freq, Q):
         Q = torch.tensor(Q)
         self.assert_grad(F.bandreject_biquad, (x, sr, central_freq, Q))
 
+    @parameterized.expand(
+        [
+            (True,),
+            (False,),
+        ]
+    )
+    def test_compute_power_spectral_density_matrix(self, use_mask):
+        torch.random.manual_seed(2434)
+        specgram = torch.rand(4, 10, 5, dtype=torch.cfloat)
+        if use_mask:
+            mask = torch.rand(10, 5)
+        else:
+            mask = None
+        self.assert_grad(F.compute_power_spectral_density_matrix, (specgram, mask))
+
+    def test_compute_mvdr_weights_souden(self):
+        torch.random.manual_seed(2434)
+        channel = 4
+        psd_speech = torch.rand(64, channel, channel, dtype=torch.cfloat)
+        psd_noise = torch.rand(64, channel, channel, dtype=torch.cfloat)
+        self.assert_grad(F.compute_mvdr_weights_souden, (psd_speech, psd_noise, 0))
+
+    def test_compute_mvdr_weights_rtf(self):
+        torch.random.manual_seed(2434)
+        channel = 4
+        rtf = torch.rand(64, channel, dtype=torch.cfloat)
+        psd_noise = torch.rand(64, channel, channel, dtype=torch.cfloat)
+        self.assert_grad(F.compute_mvdr_weights_rtf, (rtf, psd_noise, 0))
+
+    # The eigenvector can be different in different runs, expected to fail
+    @expectedFailure
```

Comment:
Reading https://pytorch.org/docs/stable/generated/torch.linalg.eigh.html, there seems no way to control the randomness. Since this is expected to fail, what value does this test provides?

Reply:
This works as a flag showing eigenvalue decomposition itself can have different results on the same matrix.

Reply:
You can normalize the rtf vector (`rtf / rtf[0].conj()`), then it should work.
If it still does not work, the eigenvalue gap should be larger, i.e. the difference between the largest and second-largest eigenvalue, e.g.:
```python
import numpy as np
channel = 4
samples = 100  # samples >> channel
data = np.random.randn(channel, samples)
cov = data @ data.T.conj()  # Hermetian matrix
eigenvalues, eigenvectors = np.linalg.eigh(cov)
eigenvalues[...] = np.random.uniform(0, 1, size=eigenvalues.shape)  # Change the eigenvalues to be between 0 and 1
eigenvalues[0] = 2  # Set first eigenvalue to be the largest
cov = np.einsum('ij,j,kj->ik', eigenvectors, eigenvalues, eigenvectors.conj())  # Revert the eigh decomposition
print(np.linalg.eigh(cov))
```

Reply:
Btw. probably the gradient for eig and eigh was derived with the constraint, that you have no gradient in the direction of the length or a phase change, i.e. the objective does not change, when you replace an eigenvector with itself multiplied with any nonzero complex number.

Reply:
IMO the returned eigenvalue can be different even when the eigenvectors are unique. One simple example is `-eigenvalue` and `-eigenvector` become another solution. 
From PyTorch side, they seem to hard-coded check it and throw a runtime error in
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L2907-L2910

Reply:
What is the source of disorder of the eigenvalues? The reason it fails is essentially multiple runs generate different results. If it is caused by values randomly initialized inside of `torch.linarg.eigh`, then wrapping the tested function with fixing random seed should work.

This is done in transform autograd test as following

https://github.com/pytorch/audio/blob/8e3c6144e88dcbbbed6dec2b47da33fac2278820/test/torchaudio_unittest/transforms/autograd_test_impl.py#L17-L27

Reply:
@mthrok  See the warning notes of [``torch.linalg.eigh``](https://pytorch.org/docs/stable/generated/torch.linalg.eigh.html).

Reply:
Before make the gradcheck work, I need to resolve the runtime error I saw in https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L2907-L2910
Here is a more detailed error message:
```python
RuntimeError: linalg_eigh_backward: The eigenvectors in the complex case are specified up to multiplication by e^{i phi}. The specified loss function depends on this quantity, so it is ill-defined.
```
It should be related with only returning the partial eigenvector. If I return the eigenvalues and eigenvectors, then the error becomes the Jacobian mismatch by gradcheck.

Reply:
If the implementation does not support gradient, then there is no point of having this test with expected failure. It simply adds the cost for testing and maintaining. Especially, since the explanation of why it fails is not on the spot, so it will be difficult for random future maintainers to decide what to do with this test.
If we want to let people know, we can simply put it down on the docstring.

Reply:
> IMO the returned eigenvalue can be different even when the eigenvectors are unique. One simple example is -eigenvalue and -eigenvector become another solution.

The eigenvalues are unique, and for a PSD matrix the eigenvalue must be non-negative. You only get an issue with duplicated eigenvalues, but then you have another issue (NaN grad). The freedom of the eigenvector is a complex scalar, in your example you selected -1. As far as I know, all implementations yield eigenvectors with l2 norm 1 (except generalized eigh), so the observable freedom is `e^{j angle}`

When you use `eigenvector / eigenvector[0].conj()` as an operation after the eigenvalue decomposition, the output will be unique. In 2017 with chainer and numpy I wrote numeric gradient testcases for the eigenvalue decomposition for beamforming, and they worked, when I considered all realization aspects (non-unique eigh output, eigh does not use all input values, ...).

Reply:
@boeddeker  I verified that using ``eigenvector / eigenvector[..., -1:].conj()`` passed the gradcheck. Thanks! In my case the eigenvalues are in ascending order, hence the normalization should rely on the last dimension instead of 0th?

Reply:
Sorry my mistake. I converted the psd matrix to real Tensor by mistake. It still fails the gradcheck now...

Reply:
> I verified that using eigenvector / eigenvector[..., -1:].conj() passed the gradcheck.

Sorry, I copied my code example from above and renamed the variable.
The correct code is
```python
...
assert rtf.ndim == 1, rtf.shape
reference_channel = 0
rtf = rtf / rtf[reference_channel].conj()
...
```
So I assumed to have a single eigenvector and normalize it by a reference channel. In this way, the eigenvectors are unique.

When I implemented the gradient check, I had issues with inputs like `psd_speech`. 
It was necessary to include something like `psd_speech = 0.5 * (psd_speech + hermitian(psd_speech))` in the checked function, otherwise the changes in the input were not correctly reflected in the output, because eigh ignores values of the matrix (e.g. all values above the diagonal aren't used). In the user code, this is not necessary, because `F.compute_power_spectral_density_matrix` takes care of this.

Reply:
Nevermind, the rtf normalization seems is there already and is True by default.